File: python/dp/coin-change.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
322. Coin Change

You are given coins of different denominations and a
total amount of money amount. Write a function to compute
the fewest number of coins that you need to make up that amount.
If that amount of money cannot be made up by any combination
of the coins, return -1.

Example 1:
coins = [1, 2, 5], amount = 11
return 3 (11 = 5 + 5 + 1)

Example 2:
coins = [2], amount = 3
return -1.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@author: houxue
@date: 2017/3/20
"""


class Solution(object):
    def coin_change(self, coins, amount):
        """
        维护一个一维动态数组dp，其中dp[i]表示钱数为i时的最小硬币数，递推式为：
            dp[i+coins[j]] = min{dp[i+coins[j]], dp[i]+1}
        即： dp[i] = min(dp[i], dp[i - coins[j]] + 1);
        其中coins[j]为第j个硬币，而i - coins[j]为钱数i减去其中一个硬币的值，
        剩余的钱数在dp数组中找到值，然后加1和当前dp数组中的值做比较，取较小的那个更新dp数组。
        :param coins:
        :param amount:
        :return:
        """
        INF = 0x7fffffff
        dp = [INF] * (amount+1)
        dp[0] = 0
        for i in range(1, amount+1):
            for coin in coins:
                if coin <= i:  # 可以放入
                    dp[i] = min(dp[i], dp[i-coin]+1)
        return dp[amount] if dp[amount] != INF else -1


if __name__ == "__main__":
    print(Solution().coin_change([1, 2, 5], 11))

File: cpp/dp/coin-change.cpp
#include <bits/stdc++.h>

using namespace std;


class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // dp[i]: 兑换总数为i的至少需要几个硬币
        vector<int> dp(amount+1, amount+1);

        dp[0] = 0;
        for (int i = 1; i <= amount; ++i) {
            for (int coin : coins) {
                if (i >= coin) {
                    dp[i] = min(dp[i], dp[i-coin]+1);
                }
            }
        }

        return dp[amount] == amount+1 ? -1 : dp[amount];
    }

    // 518. 给定不同面额的硬币和一个总金额。写出函数来计算可以凑成总金额的硬币组合数。假设每一种面额的硬币有无限个
    int change(int amount, vector<int>& coins) {
        int n = coins.size();

        // dp[i][j]: 只用前i个硬币, 兑换j金额 的组合数
        vector<vector<int>> dp(n+1, vector<int>(amount+1, 0));

        // 兑换0金额, 只有1种方式
        for (int i = 0; i <= n; ++i) {
            dp[i][0] = 1;
        }

        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= amount; ++j) {
                if (j >= coins[i-1]) {
                    // 可以放, 所以是放和不放之和
                    // 1.如果不使用这个coin, 方式同dp[i-1][j]
                    // 2.如果使用这个coin, 方式为dp[i][j-coin]
                    dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]];
                } else {
                    dp[i][j] = dp[i-1][j];
                }
            }
        }

        return dp[n][amount];
    }
};